src/baseAtomWithQuery.ts

```diff
@@ -63,92 +50,221 @@ export function baseAtomWithQuery<
+  // This atom will recalculate only once per atom reset - see returnAtom.
+  const optimisticResultAtom = atom<
+    | Promise<QueryObserverResult<TData, TError>>
+    | QueryObserverResult<TData, TError>,
+    [QueryObserverResult<TData, TError>],
+    void
+  >(
+    (get, { setSelf }) => {
+      const observer = get(observerAtom)
+      const defaultedOptions = get(defaultedOptionsAtom)
+      // Recalculate atom when query is resetted.
+      get(optimisticResultResetAtom)
 
-  return atom((get) => {
-    const observer = get(observerAtom)
-    const defaultedOptions = get(defaultedOptionsAtom)
+      const result = observer.getOptimisticResult(defaultedOptions)
 
-    const client = getQueryClient(get)
+      if (!shouldSuspend(defaultedOptions, result, false)) {
+        // Update queryResultAtom with sync result.
+        Promise.resolve(result).then(setSelf)
 
-    resetAtom.onMount = () => {
-      return () => {
-        if (observer.getCurrentResult().isError) {
-          client.resetQueries({ queryKey: observer.getCurrentQuery().queryKey })
-        }
+        return result
       }
-    }
 
-    get(resetAtom)
-    get(get(dataAtom))
+      return (
+        observer
+          .fetchOptimistic(defaultedOptions)
+          .then((succeedResult) => {
+            setSelf(succeedResult)
+
+            return succeedResult
+          })
+          // useSuspenseQuery is catching fetchOptimistic error, and triggers error boundary.
+          // Later, when error boundary is resetted, it means that useSuspenseQuery is mounting, and it's resetting the query.
+          // (therefore for useSuspenseQuery, when error boundary is active then query.state.status === 'error'? Not sure is it like this)
+          //
+          // Jotai also has to catch error, because if error would be set as atom value,
+          // then recovering from error is possible only by outside call
+          // - see https://github.com/jotaijs/jotai-tanstack-query/issues/32#issue-1684399582
+          // This is different than jotai-tanstack-query current implementation.
+          .catch(() => {
+            // Since fetchOptimistic failed, error has to be thrown, but let's do it same way as tanstack/react-query useBaseQuery.
+            // Error handling is done in returnAtom, but first, current query result must be obtained.
+            //
+            // observer.currentResult() has outdated value, and observer.updateResult() doesn't help,
+            // (I'm guessing that's due how observer.getOptimisticResult behaves with suspense = true)
+            // therefore get optimisticResult but without activating query (_optimisticResults: undefined [I think that's how it works])
+            const { _optimisticResults, ...clearedDefaultedOptions } =
+              defaultedOptions
+
+            const erroredResult = observer.getOptimisticResult(
+              clearedDefaultedOptions
+            )
```

Comment:
`_optimisticResults: undefined` is not accepted by linter/ts config, therefore do that :<

Reply:
introduced here in react query: https://github.com/TanStack/query/pull/6611 .

I'll see if I can write a test to replicate that to see if its relevant.

Reply:
if observer.getCurrentResult() is stale you can probably try tapping into queryCache like above.